### PR TITLE
Maximum sample-based training for Megatron NMT and Text Memmap based Seq2seq Pre-training

### PR DESCRIPTION
**Pull request by MaximumEntropy**
# What does this PR do ?

1. Trains Megatron-based NMT models based on maximum number of samples.
2. Added support in `text_memmap` and `csv_memmap` in Megatron encoder-decoder models (T5, BART, UL2)

**Collection**: NLP

# Usage

Add to command line

```
  model.data.data_impl=text_mmap \
  +model.data.data_impl_kwargs.newline_int=10 \
  +model.data.data_impl_kwargs.header_lines=0 \
  +model.data.data_impl_kwargs.workers=null \
  +model.data.data_impl_kwargs.sort_dataset_paths=False
```

```python
# Add a code snippet demonstrating how to use this 
```

# Before your PR is "Ready for review"
**Pre checks**:
- [ ] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [ ] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information
* Related to # (issue)

### CODE REVIEW

File: nemo/collections/nlp/data/language_modeling/megatron/gpt_dataset.py [hunk 1]
```diff
@@ -100,13 +101,13 @@ def build_train_valid_test_datasets(
     # Blend.
     blending_train_dataset = None
     if train_datasets:
-        blending_train_dataset = BlendableDataset(train_datasets, weights)
+        blending_train_dataset = BlendableDataset(train_datasets, weights, train_n)
```
[the remark this answers is not in the document]
Comment (MaximumEntropy) [reply nemo/collections/nlp/data/language_modeling/megatron/gpt_dataset.py:104]: Same as for T5/BERT.

File: examples/nlp/machine_translation/megatron_nmt_training.py [hunk 2]
```diff
@@ -145,8 +145,12 @@ def main(cfg) -> None:
             pretrained_cfg.train_ds = cfg.model.train_ds
             pretrained_cfg.train_ds.micro_batch_size = cfg.model.micro_batch_size
             pretrained_cfg.train_ds.global_batch_size = cfg.model.global_batch_size
-            pretrained_cfg.validation_ds = cfg.model.validation_ds
-            pretrained_cfg.test_ds = cfg.model.test_ds
+            if hasattr(cfg.model, 'validation_ds'):
+                pretrained_cfg.validation_ds = cfg.model.validation_ds
+            else:
+                raise AttributeError(f"No validation dataset found in config.")
```
Comment (ericharper) [inline examples/nlp/machine_translation/megatron_nmt_training.py:151]: We have to run with validation dataset?

Comment (MaximumEntropy) [reply examples/nlp/machine_translation/megatron_nmt_training.py:151]: Yeah, this is a requirement. I think it was the same with NeMo NMT as well?

File: nemo/collections/nlp/data/common/sequence_to_sequence_dataset.py [hunk 3]
```diff
@@ -146,76 +154,136 @@ def __init__(
...
+        logging.info(f'Desired number of samples : {self.max_num_samples}')
+        logging.info(f'Source Dataset Length : {len(self.src_indexed_dataset)}')
+        logging.info(f'Target Dataset Length : {len(self.tgt_indexed_dataset)}')
 
     def __len__(self):
-        return len(self.src_dataset)
+        if self.max_num_samples is None:
+            return len(self.src_indexed_dataset)
+        else:
+            return self.max_num_samples
```
Comment (ericharper) [inline nemo/collections/nlp/data/common/sequence_to_sequence_dataset.py:167]: What if `self.max_num_samples > len(self.src_indexed_dataset)` ?

File: nemo/collections/nlp/data/common/sequence_to_sequence_dataset.py [hunk 4]
```diff
@@ -146,76 +154,136 @@ def __init__(
...
+                seed=self.seed,
+                name=self.src_file_name.split('/')[-1],
+                binary_head=False,
+            )
+        else:
+            self.samples_mapping = None
+
+
+class TextMemmapSequenceToSequenceDataset(IndexedSequenceToSequenceDataset):
+    """Memory-mapped text sequence to sequence dataset. Operates on raw text files and tokenizes the text on-the-fly."""
```
Comment (ericharper) [inline nemo/collections/nlp/data/common/sequence_to_sequence_dataset.py:218]: Is there exactly one source file and one target file?

File: nemo/collections/nlp/data/common/sequence_to_sequence_dataset.py [hunk 5]
```diff
@@ -126,8 +132,8 @@ def collate_fn(self, batch):
         }
 
 
-class TextMemmapSequenceToSequenceDataset(SequenceToSequenceDataset):
-    """Sequence to Sequence Dataset in memory."""
+class IndexedSequenceToSequenceDataset(SequenceToSequenceDataset):
+    """Abstract class for TextMemmapSequenceToSequenceDataset and BinarizedMemmapSequenceToSequenceDataset."""
```
Comment (ericharper) [inline nemo/collections/nlp/data/common/sequence_to_sequence_dataset.py:136]: Please expand on the docstring here.

File: nemo/collections/nlp/data/common/sequence_to_sequence_dataset.py (hunk 4 above)
Comment (ericharper) [reply nemo/collections/nlp/data/common/sequence_to_sequence_dataset.py:218]: Please expand on the doc string here.

File: nemo/collections/nlp/data/common/sequence_to_sequence_dataset.py [hunk 6]
```diff
@@ -146,76 +154,136 @@ def __init__(
...
+        assert len(self.src_indexed_dataset) == len(
+            self.tgt_indexed_dataset
+        ), "src and tgt has different number of lines"
+        self._build_samples_mapping()
 
 
-class BinarizedMemmapSequenceToSequenceDataset(SequenceToSequenceDataset):
-    """Sequence to Sequence Dataset based on Megatron Dataset Utils."""
+class BinarizedMemmapSequenceToSequenceDataset(IndexedSequenceToSequenceDataset):
+    """Memory-mapped text sequence to sequence dataset. Operates pre-tokenized binarized data files."""
```
Comment (ericharper) [inline nemo/collections/nlp/data/common/sequence_to_sequence_dataset.py:260]: Please expand on the doc string here.

**Review by ericharper — CHANGES_REQUESTED** [review]
Overall looks good. Main request is to expand on the doc strings for the datasets. When to use them. What do the source and target files need to look like. Can you use more than source and target file. Etc.

File: nemo/collections/nlp/data/common/sequence_to_sequence_dataset.py (hunk 3 above)
Comment (MaximumEntropy) [reply nemo/collections/nlp/data/common/sequence_to_sequence_dataset.py:167]: That's fine. `build_samples_mapping()` will then "upsample" your data to the provided `max_num_samples` amount to turn everything into single epoch training like we do for pre-training.

File: nemo/collections/nlp/data/common/sequence_to_sequence_dataset.py (hunk 4 above)
Comment (MaximumEntropy) [reply nemo/collections/nlp/data/common/sequence_to_sequence_dataset.py:218]: Done. Added a docstring to the constructor that explains each input arg to the class.

File: nemo/collections/nlp/data/common/sequence_to_sequence_dataset.py (hunk 6 above)
Comment (MaximumEntropy) [reply nemo/collections/nlp/data/common/sequence_to_sequence_dataset.py:260]: Added a docstring to the constructor here as well.

File: nemo/collections/nlp/data/common/sequence_to_sequence_dataset.py (hunk 5 above)
Comment (MaximumEntropy) [reply nemo/collections/nlp/data/common/sequence_to_sequence_dataset.py]: Add a docstring here as well.

**Review by ericharper — APPROVED** [review]
LGTM. Thanks for the changes!

**Review by ericharper — CHANGES_REQUESTED** [review]
Blocking while we figure out CI issue.

**Review by ericharper — APPROVED** [review]
Unblocking.

**Review by ericharper — DISMISSED** [review]
Thanks!

**Review by ericharper — APPROVED** [review]
LGTM. Thanks!

**Pull request merged**